tools/protoc-gen-docs/baseDesc.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+)
+
+// coreDesc is an interface abstracting the abilities shared by all descriptors
+type coreDesc interface {
+	fileDesc() *fileDescriptor
+	qualifiedName() []string
+	isHidden() bool
+	isExperimental() bool
+	location() *descriptor.SourceCodeInfo_Location
+}
+
+// The common data for every descriptor in the model. This implements the coreDesc interface.
+type baseDesc struct {
+	loc          *descriptor.SourceCodeInfo_Location
+	hidden       bool
+	experimental bool
+	file         *fileDescriptor
+	name         []string
+}
+
+func newBaseDesc(file *fileDescriptor, path pathVector, qualifiedName []string) baseDesc {
+	loc := file.find(path)
+	experimental := false
+	com := ""
+
+	if loc != nil {
+		com = loc.GetLeadingComments()
```

Comment:
Is there a reason to distinguish between leading and trailing comments... can we just merge the two ? or just pick leading if there, else trailing.. This code then then be simplified by writing the $experimental  related code just once.

Reply:
It makes it so you can add the annotation above the element or to the right of the element, which is awfully convenient.